File: bridge/lib/web/bridge_achievements.js
let js_bridge_achievements = {
    js_bridge_achievements_unlock: function (handler, id, onSuccess, onFailure) {
        bridge.achievements.unlock(UTF8ToString(id))
            .then(result => {
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 0, packToJson(result));
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 1, packToJson(error));
            })
    },

    js_bridge_achievements_getAchievements: function (handler, onSuccess, onFailure) {
        bridge.achievements.getAchievements()
            .then(result => {
                const jsonString = JSON.stringify(result);
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 0, packToJson(jsonString));
            })
            .catch(error => {
                {{{ makeDynCall('viiii', 'handler') }}} (onSuccess, onFailure, 1, packToJson(error));
            })
    }
}

mergeInto(LibraryManager.library, js_bridge_achievements);

File: bridge/src/bridge_achievements.cpp
#if defined(DM_PLATFORM_HTML5)
#include "bridge_achievements.h"
#include "bridge.h"
#include "bridge_helper.h"

int bridge::achievements::unlock(lua_State* L) {
    return makeCallbackWithString(L, js_bridge_achievements_unlock, true);
}

int bridge::achievements::getAchievements(lua_State* L) {
    return makeCallback(L, js_bridge_achievements_getAchievements, true);
}

#endif